Support azure and abfss in LLM config

gangsf opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

We have verified in the template test https://console.anyscale-staging.com/cld_5nnv7pt2jn2312x2e5v72z53n2/prj_y8syktydl7ltabhz5axdelwnce/workspaces/expwrk_7gnc5p1a6yvv77i74bdh9h52q5?command-history-section=application_logs&workspace-tab=overview

with the image built from this PR

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Add Azure Blob Storage and ABFSS support for path detection, mirror configs, filesystem operations, and LoRA config validation, with comprehensive tests.
> 
> - **Cloud utils**:
>   - Extend `is_remote_path` to recognize `abfss://` and `azure://`.
>   - Update `CloudMirrorConfig`/`LoraMirrorConfig` URI validation messages; add `storage_type` values `abfss`/`azure`; parse Azure container names (`container@account`).
>   - `CloudFileSystem.get_fs_and_path` now supports `abfss://` and `azure://`; add `_create_azure_filesystem` and `_create_abfss_filesystem` using `adlfs` + `DefaultAzureCredential`; limit anonymous access parsing to S3/GCS.
> - **Serve config**:
>   - Broaden `LoraConfig.dynamic_lora_loading_path` validation to accept `abfss://` and `azure://`.
> - **Tests**:
>   - Add tests for Azure/ABFSS path detection, config validation/parsing, filesystem creation, file ops, URI validation, and import-error handling.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 4ebc5bbcc00331a4ba563b40d869453d32fe7543. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

Commit:
Support azure and abfss in LLM config

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Make azure and abfss distinct

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

nrghosh reviewed (commented):
+1 for test - otherwise lgtm

Commit:
Add Azure/ABFSS support to Ray LLM cloud utilities

- Add _create_abfss_filesystem() and _create_azure_filesystem() helper methods to CloudFileSystem
- Update get_fs_and_path() to detect and handle abfss:// and azure:// URIs
- Integrate adlfs.AzureBlobFileSystem with PyArrow's PyFileSystem for compatibility
- Update is_remote_path() to recognize abfss:// and azure:// protocols
- Add storage_type property support for 'abfss' and 'azure' in CloudMirrorConfig and LoraMirrorConfig
- Add comprehensive unit tests for Azure/ABFSS functionality including:
  - URI validation and parsing tests
  - Filesystem creation and error handling tests
  - Configuration class tests for Azure URIs
  - Integration tests for CloudFileSystem methods with ABFSS

Requires: pip install adlfs azure-identity
Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

gangsf commented:
> (I see this was just opened, not sure if still WIP) but can an integration test be added?

Kevin is setting up the AKS env in CI. Once that is done, we can add some integration test.

Commit:
Fix LoraMirrorConfig test failures by adding required max_total_tokens field

- Added max_total_tokens=1000 to all LoraMirrorConfig test instantiations
- The field appears to be required despite being marked as Optional[int]
- This fixes the 4 failing test cases in TestLoraMirrorConfig class
- All Azure/ABFSS functionality tests now pass successfully

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Fix bucket_name property for ABFSS and Azure URIs

- Updated bucket_name property in LoraMirrorConfig to correctly extract container name
- For ABFSS/Azure URIs (container@account.domain), extract only 'container' part
- Maintains backward compatibility with S3/GCS URIs (bucket/path format)
- Fixes test failures in TestLoraMirrorConfig for Azure/ABFSS bucket name parsing
- All URI formats now correctly return just the bucket/container name

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

eicherseiji commented on python/ray/llm/_internal/common/utils/cloud_utils.py:
```diff
@@ -159,11 +176,158 @@ def get_fs_and_path(object_uri: str) -> Tuple[pa_fs.FileSystem, str]:
         return fs, path
 
+    @staticmethod
+    def _create_abfss_filesystem(object_uri: str) -> Tuple[pa_fs.FileSystem, str]:
```
Seems like there's a lot of duplicated code between this and `_create_azure_filesystem`, any reason we can't unify them?

eicherseiji commented on python/ray/llm/_internal/common/utils/cloud_utils.py:
```diff
@@ -37,7 +37,12 @@ def is_remote_path(path: str) -> bool:
         True if the path is a remote path, False otherwise.
     """
-    return path.startswith("s3://") or path.startswith("gs://")
+    return (
+        path.startswith("s3://")
+        or path.startswith("gs://")
+        or path.startswith("abfss://")
+        or path.startswith("azure://")
+    )
```
+1

eicherseiji reviewed (commented):
Thanks @gangsf. Is there a test somewhere that actually hits non-mocked Azure/abfss?

gangsf commented:
> Thanks @gangsf. Is there a test somewhere that actually hits non-mocked Azure/abfss?

We have verified in the template test https://console.anyscale-staging.com/cld_5nnv7pt2jn2312x2e5v72z53n2/prj_y8syktydl7ltabhz5axdelwnce/workspaces/expwrk_7gnc5p1a6yvv77i74bdh9h52q5?command-history-section=application_logs&workspace-tab=overview

with the image built from this PR

Commit:
Unify _create_abfss_filesystem and _create_azure_filesystem

- Refactor duplicate code into a single _create_azure_filesystem function
- Support both abfss:// and azure:// schemes in the unified function
- Keep _create_abfss_filesystem as a backward-compatible wrapper
- Reduce code duplication and improve maintainability

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

Commit:
Apply formatting fixes to cloud_utils.py

- Reformat boolean condition for better readability
- Remove trailing whitespace
- Applied via scripts/format.sh

Signed-off-by: Gang Zhao <gang@gang-JQ62HD2C37.local>

nrghosh reviewed (commented):
+1 for having an actual integration test that hits the filesystem vs. testing on a static image

gangsf commented:
> > We have verified in the template test https://console.anyscale-staging.com/cld_5nnv7pt2jn2312x2e5v72z53n2/prj_y8syktydl7ltabhz5axdelwnce/workspaces/expwrk_7gnc5p1a6yvv77i74bdh9h52q5?command-history-section=application_logs&workspace-tab=overview
> > with the image built from this PR
> 
> @gangsf I mean, is there a test that verifies this behavior live in CI? If not, can we add one?

Kevin is working on the CI setup for AKS. We can add one after that

nrghosh reviewed (approved):
Sounds good

gangsf committed:
Merge branch 'master' into gang/abfss_support_in_llm
